### PR TITLE
Scheduled publishing

### PR DESCRIPTION
**Pull request by tomkins**
Currently needs tests, but I'd like some feedback to avoid any surprises.

To add this to the admin for a glitterable object, do something like:

```python
from glitter.publisher.admin import ActionInline

class PageAdmin(GlitterAdminMixin, DjangoMpttAdmin, MPTTModelAdmin):
    inlines = [ActionInline]
```

This means scheduled publishings are added/edited as part of the Django admin object edit. Just add a new inline for what version you want to be published, and when you want it to happen.

Once the task is run, the object disappears to keep the list of scheduled actions as small as possible. However we add something to the Django admin history log so users know why something has been published/unpublished.

To keep this flexible - we can work with 2 modes of running this:

First option is crontab - just add the task to run every minute. This doesn't give accuracy to the second, but everything runs in order.

Second option is celery. When saving an object it schedules a task to run at the exact time. We can also add a recurring task for celery beat incase of tasks not being run due to any problems, it'll go through all tasks which should've been processed and actions them.

There's a few bits where I need to harden it from disappearing objects - but that shouldn't be too difficult.

### CODE REVIEW

File: glitter/publisher/utils.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def process_actions(action_ids=None):
+    """
+    Process actions in the publishing schedule.
+
+    Returns the number of actions processed.
+    """
+    actions_taken = 0
+    action_list = PublishAction.objects.prefetch_related(
```
Comment (zatan) [inline glitter/publisher/utils.py]: Is this helps anyway by doing prefetch_related on the `content_object`  I find out that SQL still remains the same if I use `prefetch_related` ?

Comment (tomkins) [inline glitter/publisher/utils.py]: This seems to work fine for me.

You'll need to add the inline onto 2 different models, then this statement does something similar to following:

- Select all publishactions
- Select all pages for publishactions where content type is page
- Select all news posts for publishactions where content type is news post
- etc.

File: glitter/publisher/admin.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from .models import PublishAction
+
+
+class ActionInline(GenericStackedInline):
+    model = PublishAction
+    fields = ('scheduled_time', 'publish_version')
+    extra = 0
+
+    def get_formset(self, request, obj=None, form=None, **kwargs):
+        class VersionForm(forms.ModelForm):
```
Comment (zatan) [inline glitter/publisher/admin.py]: I think we discuss this previously but I prefer going this way:

```        formset = super(ActionInline, self).get_formset(request, obj, **kwargs)
        formset.form.base_fields['publish_version'].widget = forms.widgets.Select(
            choices=object_version_choices(obj=obj)
        )
        formset.request = request
        
        return formset
```
It's just 4 lines also you need to set at the top `formset = ActionInlineFormset` and create form set with the `save_new` and `save_existing` methods hope this helps.

File: glitter/publisher/apps.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from __future__ import unicode_literals
+
+from django.apps import AppConfig
+
+
+class PublisherConfig(AppConfig):
```
Comment (zatan) [inline glitter/publisher/apps.py]: Is there any specific reason why this define? As I think I haven't defined `apps.py` as I felt no need for it.

File: glitter/publisher/management/commands/publish_task.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+from django.core.management.base import BaseCommand
+
+from glitter.publisher.utils import process_actions
+
+
+class Command(BaseCommand):
+    help = 'Process scheduled publishing for Glitter objects'
+
+    requires_system_checks = False
```
Comment (zatan) [inline glitter/publisher/management/commands/publish_task.py]: +1

File: glitter/publisher/apps.py (hunk 3 above)
Comment (tomkins) [inline glitter/publisher/apps.py]: Mostly to keep the `glitter` prefix to avoid naming problems.

**Comment by zatan** [comment]
Is this should have enabled/disable flag for the core PageAdmin class? To prevent overriding PageAdmin if this should be enabled.

File: glitter/publisher/models.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        if self.publish_version == self.UNPUBLISH_CHOICE:
+            return 'Unpublish at {}'.format(self.scheduled_time)
+        else:
+            return 'Publish version {} at {}'.format(self.publish_version, self.scheduled_time)
+
+    def save(self, *args, **kwargs):
+        super(PublishAction, self).save(*args, **kwargs)
+
+        # Add a task
+        if self.can_schedule_task():
```
Comment (zatan) [inline glitter/publisher/models.py]: I think we need some more logic here to prevent adding same tasks?

Comment (tomkins) [inline glitter/publisher/models.py]: Not really.

The publish/process code is fairly defensive against multiple tasks. If the object changes date, then the queryset in `process_actions` won't see/select it. If the object is deleted as the task was already completed earlier - it won't be able to see/select it again.

**Comment by zatan** [comment]
One little thing to add the validation to make sure the date they entered is not in the past.

**Pull request merged**